validator/client/src/test/java/tech/pegasys/teku/validator/client/loader/ExternalValidatorSourceTest.java

```diff
@@ -116,4 +128,83 @@ void shouldThrowExceptionWhenDeleteValidator() {
   void shouldSayFalseToAddValidators() {
     assertThat(validatorSource.canUpdateValidators()).isFalse();
   }
+
+  @Test
+  void shouldAddValidator_getsConfigUrlWhenNotProvided(@TempDir Path tempDir) throws IOException {
+    BLSPublicKey publicKey = dataStructureUtil.randomPublicKey();
+    final AddValidatorResult result =
+        getResultFromAddingValidator(tempDir, publicKey, Optional.empty());
+    assertThat(result.getResult().getImportStatus()).isEqualTo(ImportStatus.IMPORTED);
+    assertThat(result.getSigner()).isNotEmpty();
+    assertThat(result.getSigner().get().getSigningServiceUrl()).isNotEmpty();
+    assertThat(result.getSigner().get().getSigningServiceUrl().get())
+        .isEqualTo(config.getValidatorExternalSignerUrl());
+
+    String fileName = publicKey.toBytesCompressed().toUnprefixedHexString() + ".json";
+    Path path =
+        ValidatorClientService.getManagedRemoteKeyPath(dataDirLayout(tempDir)).resolve(fileName);
+    assertThat(path.toFile()).exists();
+    assertThat(Files.toString(path.toFile(), Charsets.UTF_8))
+        .isEqualTo("{\"pubkey\":\"" + publicKey + "\"}");
+  }
+
+  @Test
+  void shouldAddValidator_getsUrlWhenProvided(@TempDir Path tempDir) throws IOException {
+    BLSPublicKey publicKey = dataStructureUtil.randomPublicKey();
+    URL url = new URL("http://host.com");
+    final AddValidatorResult result =
+        getResultFromAddingValidator(tempDir, publicKey, Optional.of(url));
+    assertThat(result.getResult().getImportStatus()).isEqualTo(ImportStatus.IMPORTED);
+    assertThat(result.getSigner()).isNotEmpty();
+    assertThat(result.getSigner().get().getSigningServiceUrl()).isNotEmpty();
+    assertThat(result.getSigner().get().getSigningServiceUrl().get()).isEqualTo(url);
+
+    // Check contents of the file
+    String fileName = publicKey.toBytesCompressed().toUnprefixedHexString() + ".json";
+    Path path =
+        ValidatorClientService.getManagedRemoteKeyPath(dataDirLayout(tempDir)).resolve(fileName);
+    assertThat(path.toFile()).exists();
+    assertThat(Files.toString(path.toFile(), Charsets.UTF_8))
+        .isEqualTo("{\"pubkey\":\"" + publicKey + "\",\"url\":\"http://host.com\"}");
+  }
+
+  @Test
+  void shouldDetectDuplicatesOnAddValidator(@TempDir Path tempDir) throws IOException {
+    BLSPublicKey publicKey = dataStructureUtil.randomPublicKey();
+    final AddValidatorResult result1 =
+        getResultFromAddingValidator(tempDir, publicKey, Optional.empty());
+    assertThat(result1.getResult().getImportStatus()).isEqualTo(ImportStatus.IMPORTED);
+    assertThat(result1.getSigner()).isNotEmpty();
+    assertThat(result1.getSigner().get().getSigningServiceUrl()).isNotEmpty();
+    assertThat(result1.getSigner().get().getSigningServiceUrl().get())
+        .isEqualTo(config.getValidatorExternalSignerUrl());
```

Comment:
i think this code is heavily repeated, could potentially extract it to a check function that takes a pubkey

Reply:
or just check status once its been verified that the values are right in another test...
or it may just mean the validations are mostly inside the getter fn you're using